fog/view/server/src/fog_view_shard_responses_processor.rs

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use crate::error::RouterServerError;
+
+use mc_attest_api::attest;
+use mc_fog_api::{view::MultiViewStoreQueryResponse, view_grpc::FogViewApiClient};
+use mc_fog_uri::FogViewUri;
+use std::{str::FromStr, sync::Arc};
+
+/// The result of processing the MultiViewStoreQueryResponse from each Fog View
+/// Shard.
+pub struct ProcessedShardResponseData {
+    /// gRPC clients for Shards that need to be retried for a successful
+    /// response.
+    pub shard_clients_for_retry: Vec<Arc<FogViewApiClient>>,
+
+    /// Uris for *individual* Fog View Stores that need to be authenticated with
+    /// by the Fog Router.
+    pub view_store_uris_for_authentication: Vec<FogViewUri>,
+
+    /// New, successfully processed query responses.
+    pub new_query_responses: Vec<attest::Message>,
+}
+
+impl ProcessedShardResponseData {
+    pub fn new(
+        shard_clients_for_retry: Vec<Arc<FogViewApiClient>>,
+        view_store_uris_for_authentication: Vec<FogViewUri>,
+        new_query_responses: Vec<attest::Message>,
+    ) -> Self {
+        ProcessedShardResponseData {
+            shard_clients_for_retry,
+            view_store_uris_for_authentication,
+            new_query_responses,
+        }
+    }
+}
+
+/// Processes the MultiViewStoreQueryResponses returned by each Fog View Shard.
+pub fn process_shard_responses(
+    clients_and_responses: Vec<(Arc<FogViewApiClient>, MultiViewStoreQueryResponse)>,
+) -> Result<ProcessedShardResponseData, RouterServerError> {
+    let mut shard_clients_for_retry = Vec::new();
+    let mut view_store_uris_for_authentication = Vec::new();
+    let mut new_query_responses = Vec::new();
+
+    for (shard_client, mut response) in clients_and_responses {
+        // We did not receive a query_response for this shard.Therefore, we need to:
```

Comment:
I can't decide if this is mixing up the "response" with "how to handle the response".

Perhaps this should just return the failed clients and the failure handler should be the one decomposing to authenticate and retry.

Looking further it seems that the uri is tied to the store, while the shard is the front end (load balancer in current design)...

I'm at a loss for any good suggestion here, leaving comment in case one more knowledgeable has an idea.